pkg/diff/checkpoint.go

```diff
@@ -0,0 +1,256 @@
+		}
+
+		if cfgHash.Valid {
+			if configHash != cfgHash.String {
+				return false, nil
+			}
+		}
+
+		if state.Valid {
+			// if state is not checked, the chunk info maybe not exists, so just return false
+			if state.String == successState || state.String == notCheckedState {
+				return false, nil
+			}
+		}
+
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func initSummaryInfo(ctx context.Context, db *sql.DB, schema, table string, configHash string) error {
+	sql := "REPLACE INTO `sync_diff_inspector`.`table_summary`(`schema`, `table`, `state`, `config_hash`, `update_time`) VALUES(?, ?, ?, ?, ?)"
+	err := dbutil.ExecSQLWithRetry(ctx, db, sql, schema, table, notCheckedState, configHash, time.Now())
+	if err != nil {
+		log.Error("save summary info failed", zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+func loadChunksInfo(ctx context.Context, db *sql.DB, instanceID, schema, table string) ([]*ChunkRange, error) {
+	chunks := make([]*ChunkRange, 0, 100)
+
+	ctx, cancel := context.WithTimeout(context.Background(), dbutil.DefaultTimeout)
+	defer cancel()
+	query := "SELECT `chunk_str` FROM `sync_diff_inspector`.`chunk` WHERE `instance_id` = ? AND `schema` = ? AND `table` = ?"
+	rows, err := db.QueryContext(ctx, query, instanceID, schema, table)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		fields, err1 := dbutil.ScanRow(rows)
+		if err1 != nil {
+			return nil, errors.Trace(err1)
+		}
+
+		chunkStr := fields["chunk_str"].Data
+		chunk := new(ChunkRange)
+		err := json.Unmarshal(chunkStr, &chunk)
+		if err != nil {
+			return nil, err
+		}
+		chunks = append(chunks, chunk)
+	}
+
+	return chunks, nil
+}
+
+func updateSummaryInfo(ctx context.Context, db *sql.DB, instanceID, schema, table string) error {
+	ctx, cancel := context.WithTimeout(ctx, dbutil.DefaultTimeout)
+	defer cancel()
+
+	query := "SELECT `state`, COUNT(*) FROM `sync_diff_inspector`.`chunk` WHERE `instance_id` = ? AND `schema` = ? AND `table` = ? GROUP BY `state` ;"
+	rows, err := db.QueryContext(ctx, query, instanceID, schema, table)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer rows.Close()
+
+	var total, successNum, failedNum, ignoreNum int64
+	var chunkState sql.NullString
+	var num sql.NullInt64
+	for rows.Next() {
+		err1 := rows.Scan(&chunkState, &num)
+		if err1 != nil {
+			return errors.Trace(err1)
+		}
+
+		if !chunkState.Valid || !num.Valid {
+			continue
+		}
+
+		total += num.Int64
+		switch chunkState.String {
+		case successState:
+			successNum = num.Int64
+		case failedState, errorState:
+			failedNum += num.Int64
+		case ignoreState:
+			ignoreNum += num.Int64
+		case notCheckedState, checkingState:
+		}
+	}
+
+	log.Info("summary info", zap.String("instance_id", instanceID), zap.String("schema", schema), zap.String("table", table), zap.Int64("chunk num", total), zap.Int64("success num", successNum), zap.Int64("failed num", failedNum), zap.Int64("ignore num", ignoreNum))
```

Comment:
how is the performance of this log library?

Reply:
I haven't do any test , I will ask the author of pingcap/log

Reply:
https://github.com/uber-go/zap#performance